### PR TITLE
Add pivot wider

### LINKED ISSUE #543: Add pivot_wider()

### PR DESCRIPTION
**Pull request by mgirlich**
Closes #543.

The main difference to local data frames is that `values_fn` is mandatory (default `max()`). This is because the data needs to be grouped and then an aggregate function is required. If observations are uniquely identified by the combination of `id_cols` and `value` column this should not matter. One could add a (optional?) check for this (something like `collect(count(data, id_cols))`) but I'm not sure whether this is worth it.

~~The tests are not yet adapted properly but I will do so in the next days.~~
Edit: I had a go at the tests.

### CODE REVIEW

File: R/verb-pivot-wider.R [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+#' and `value` column uniquely identify an observation.
+#' Mind that you also do not get a warning if an observation is not uniquely
+#' identified.
+#'
+#' The translation to SQL code basically works as follows:
+#'
+#' 1. Get unique keys in `names_from` column.
+#' 2. For each key value generate an expression of the form:
+#'
+#' ```sql
```
Comment (hadley) [inline R/verb-pivot-wider.R:21]: Should be indented, I think to stay inside the list?

File: R/verb-pivot-wider.R [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+#' value_fn(
+#'   CASE WHEN (`names from column` == `key value`) THEN (`value column`)
+#' ) AS `output column`
+#' ```
+#'
+#' 3. Group data by id columns.
+#' 4. Summarise the grouped data with the expressions from step 2.
+#'
+#' For example the following R code
```
Comment (hadley) [inline R/verb-pivot-wider.R:30]: Move this to the examples?

File: R/verb-pivot-wider.R [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+  }
+
+  if (is.list(x)) {
+    (length(x) == 1) && !have_name(x)
+  } else {
+    length(x) == 1
+  }
+}
+
+# TODO remove when `pivot_longer.tbl_lazy()` is merged
```
Comment (hadley) [inline R/verb-pivot-wider.R:253]: ✔️

File: tests/testthat/_snaps/verb-pivot-wider.md [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    Output
+      <SQL>
+      SELECT `a`, SUM(CASE WHEN (`key` = 'x') THEN (`val`) END) AS `x`
+      FROM `df`
+      GROUP BY `a`
+
+# values_fn cannot be NULL
+
+    Code
+      expect_error(dbplyr_pivot_wider_spec(df, spec1, values_fn = NULL))
```
Comment (hadley) [inline tests/testthat/_snaps/verb-pivot-wider.md:34]: Needs `()` around `expect_error()` to see the actual error text here.

File: tests/testthat/test-verb-pivot-wider.R [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+
+  expect_snapshot(
+    lazy_frame(a = 1:2, key = c("x", "y"), val = 1:2) %>%
+      dbplyr_pivot_wider_spec(spec)
+  )
+})
+
+test_that("error when overwriting existing column", {
+  # TODO
+  skip("not yet implemented")
```
Comment (hadley) [inline tests/testthat/test-verb-pivot-wider.R:56]: Do you want to have a go at this now?

File: tests/testthat/test-verb-pivot-wider.R [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+      "Bob",   20,      NA,
+      "Sam",   10,     1.5
+    )
+  )
+})
+
+
+# non-unqiue keys ---------------------------------------------------------
+
+# TODO check error if `values_fn = NULL`
```
Comment (hadley) [inline tests/testthat/test-verb-pivot-wider.R:137]: Done now?

File: tests/testthat/test-verb-pivot-wider.R [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+  df <- lazy_frame(a = c(1, 1, 2), key = c("x", "x", "x"), val = c(1, 10, 100))
+
+  expect_snapshot(dbplyr_pivot_wider_spec(df, spec1, values_fn = sum))
+})
+
+test_that("values_fn cannot be NULL", {
+  df <- lazy_frame(a = 1, key = "x", val = 1)
+
+  expect_snapshot(
+    expect_error(dbplyr_pivot_wider_spec(df, spec1, values_fn = NULL))
```
Comment (hadley) [inline tests/testthat/test-verb-pivot-wider.R:149]: For this special case, you can also use `expect_snapshot_error()`.

File: tests/testthat/_snaps/verb-pivot-wider.md [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+# can fill in missing cells
+
+    Code
+      dbplyr_pivot_wider_spec(df_lazy, spec, values_fill = 0)
+    Output
+      <SQL>
+      SELECT `g`, `name`, `value`, MAX(CASE WHEN (`key` = 'x') THEN (`val`) WHEN NOT(`key` = 'x') THEN (0.0) END) AS `x`, MAX(CASE WHEN (`key` = 'y') THEN (`val`) WHEN NOT(`key` = 'y') THEN (0.0) END) AS `y`
```
Comment (hadley) [inline tests/testthat/_snaps/verb-pivot-wider.md:42]: I wonder if this should be `COALESCE(MAX(CASE WHEN ...), 0.0)`? What should happen if `fill = 1` and `values_fn = sum` and there are multiple missing values?

**Review by hadley — COMMENTED** [review]
Looking good!

File: tests/testthat/_snaps/verb-pivot-wider.md (hunk 8 above)
Comment (mgirlich) [reply tests/testthat/_snaps/verb-pivot-wider.md:42]: I am not sure I understand correctly what you mean but I think it is explicitly wanted to not use `coalesce()`. At least I adapted this test from `tidyr`

```r
test_that("values_fill only affects missing cells", {
  df <- memdb_frame(g = c(1, 2), name = c("x", "y"), value = c(1, NA))
  out <- tidyr::pivot_wider(df, values_fill = 0) %>%
    collect()
  expect_equal(out$y, c(0, NA))
})
```

Or here directly with `tidyr`:

``` r
library(tidyr)
df <- tibble(g = c(1, 2, 2), name = c("x", "y", "y"), value = c(1, NA, 2))
df
#> # A tibble: 3 x 3
#>       g name  value
#>   <dbl> <chr> <dbl>
#> 1     1 x         1
#> 2     2 y        NA
#> 3     2 y         2

pivot_wider(df, values_fill = 1, values_fn = sum)
#> # A tibble: 2 x 3
#>       g     x     y
#>   <dbl> <dbl> <dbl>
#> 1     1     1     1
#> 2     2     1    NA

pivot_wider(df, values_fill = 1, values_fn = function(x) sum(x, na.rm = TRUE))
#> # A tibble: 2 x 3
#>       g     x     y
#>   <dbl> <dbl> <dbl>
#> 1     1     1     1
#> 2     2     1     2
```

So, `values_fill` actually ignores `NA`. Does this answer your question or did I misunderstand you?

Comment (hadley) [reply tests/testthat/_snaps/verb-pivot-wider.md:48]: Ok, cool, that puts my mind at rest.

**Pull request merged**